common/argparse_test.cpp

```diff
@@ -0,0 +1,853 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "common/argparse.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "llvm/Support/FormatVariadic.h"
+#include "testing/util/test_raw_ostream.h"
+
+namespace Carbon::Testing {
+namespace {
+
+using ::testing::ElementsAre;
+using ::testing::Eq;
+using ::testing::HasSubstr;
+using ::testing::Not;
+using ::testing::StrEq;
+
+constexpr Args::CommandInfo TestCommandInfo = {
+    .name = "test",
+    .help = "A test command.",
+    .help_epilog = "TODO",
+};
+
+enum class TestEnum {
+  Val1,
+  Val2,
+};
+
+enum class TestSubcommand {
+  Sub1,
+  Sub2,
+};
+
+TEST(ArgParserTest, BasicCommand) {
+  bool flag = false;
+  ssize_t integer_option = -1;
+  llvm::StringRef string_option = "";
+  auto parse = [&](llvm::ArrayRef<llvm::StringRef> args) {
+    return Args::Parse(
+        args, llvm::errs(), llvm::errs(), TestCommandInfo, [&](auto& b) {
+          b.AddFlag({.name = "flag"}, [&](auto& arg_b) { arg_b.Set(&flag); });
+          b.AddIntegerOption({.name = "option1"},
+                             [&](auto& arg_b) { arg_b.Set(&integer_option); });
+          b.AddStringOption({.name = "option2"},
+                            [&](auto& arg_b) { arg_b.Set(&string_option); });
+          b.Do([] {});
+        });
+  };
```

Comment:
This may be personal preference, but the structuring of this with nested lambdas for callbacks instead of an object that has flags added then is run (which is how Python's argparse is set up) feels like it yields a more difficult to read end result.